### PR TITLE
feat: Add API to get sub menus in MenuBarElement

### LINKED ISSUE #3471: MenuBarElement should have API to get sub menues
### Describe your motivation

Current API is minimal and it requires expert knowledge on MenuBar internal implementation to operate sub menues.

### Describe the solution you'd like

At least basic API is needed to get sub menues similar that are already implemented as internal helpers in our integration tests.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by TatuLund**
Fixes #3471

### CODE REVIEW

File: vaadin-menu-bar-flow-parent/vaadin-menu-bar-testbench/src/main/java/com/vaadin/flow/component/menubar/testbench/MenuBarElement.java [hunk 1]
```diff
@@ -64,4 +70,55 @@ private boolean isVisible(TestBenchElement element) {
...
+        return (TestBenchElement) getDriver()
+                .findElement(By.tagName(OVERLAY_TAG));
+    }
+
+    /**
+     * Get all the open sub menu overlay elements.
+     *
+     * @return List of TestBenchElements representing currently open sub menues.
+     */
+    public List<TestBenchElement> getAllSubMenues() {
```
Comment (web-padawan) [inline vaadin-menu-bar-flow-parent/vaadin-menu-bar-testbench/src/main/java/com/vaadin/flow/component/menubar/testbench/MenuBarElement.java:111]:
```suggestion
    public List<TestBenchElement> getAllSubMenus() {
```

File: vaadin-menu-bar-flow-parent/vaadin-menu-bar-testbench/src/main/java/com/vaadin/flow/component/menubar/testbench/MenuBarElement.java [hunk 2]
```diff
@@ -64,4 +70,55 @@ private boolean isVisible(TestBenchElement element) {
...
+    public TestBenchElement getSubMenu() {
+        waitForSubMenu();
+        return (TestBenchElement) getDriver()
+                .findElement(By.tagName(OVERLAY_TAG));
+    }
+
+    /**
+     * Get all the open sub menu overlay elements.
+     *
+     * @return List of TestBenchElements representing currently open sub menues.
```
Comment (web-padawan) [inline vaadin-menu-bar-flow-parent/vaadin-menu-bar-testbench/src/main/java/com/vaadin/flow/component/menubar/testbench/MenuBarElement.java:109]:
```suggestion
     * @return List of TestBenchElements representing currently open sub menus.
```

File: vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java [hunk 3]
```diff
@@ -72,7 +72,7 @@ public void clickRootButton_hoverOnParentItem_subSubMenuRenders() {
         openSubSubMenu();
 
         String[] menuItemContents = getOverlayMenuItemContents(
-                getAllOverlays().get(1));
+                menuBar.getAllSubMenues().get(1));
```
Comment (web-padawan) [inline vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java:75]:
```suggestion
                menuBar.getAllSubMenus().get(1));
```

File: vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java [hunk 4]
```diff
@@ -81,42 +81,44 @@ public void clickRootButton_hoverOnParentItem_subSubMenuRenders() {
...
+        menuBar.getSubMenuItems().get(0).click();
         assertMessage("clicked sub item 1");
     }
 
     @Test
     public void openSubSubMenu_clickCheckableItem_checkableStateChanges() {
         openSubSubMenu();
 
-        getOverlayMenuItems(getAllOverlays().get(1)).get(1).click();
+        menuBar.getSubMenuItems(menuBar.getAllSubMenues().get(1)).get(1)
```
Comment (web-padawan) [inline vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java:92]:
```suggestion
        menuBar.getSubMenuItems(menuBar.getAllSubMenus().get(1)).get(1)
```

File: vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java [hunk 5]
```diff
@@ -81,42 +81,44 @@ public void clickRootButton_hoverOnParentItem_subSubMenuRenders() {
...
         assertMessage("true");
 
         menuBar.$("vaadin-menu-bar-button").first().click();
-        hoverOn(getOverlayMenuItems().get(1));
+        hoverOn(menuBar.getSubMenuItems().get(1));
 
-        waitUntil(driver -> getAllOverlays().size() == 2);
-        TestBenchElement checkableItem = getOverlayMenuItems(
-                getAllOverlays().get(1)).get(1);
+        waitUntil(driver -> menuBar.getAllSubMenues().size() == 2);
```
Comment (web-padawan) [inline vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java:101]:
```suggestion
        waitUntil(driver -> menuBar.getAllSubMenus().size() == 2);
```

File: vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java [hunk 6]
```diff
@@ -81,42 +81,44 @@ public void clickRootButton_hoverOnParentItem_subSubMenuRenders() {
...
+        TestBenchElement checkableItem = menuBar
+                .getSubMenuItems(menuBar.getAllSubMenues().get(1)).get(1);
```
Comment (web-padawan) [inline vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java:103]:
```suggestion
                .getSubMenuItems(menuBar.getAllSubMenus().get(1)).get(1);
```

File: vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java [hunk 7]
```diff
@@ -81,42 +81,44 @@ public void clickRootButton_hoverOnParentItem_subSubMenuRenders() {
...
         Assert.assertTrue(checkableItem.hasAttribute("menu-item-checked"));
 
         checkableItem.click();
         verifyClosed();
         assertMessage("false");
 
         openSubSubMenu();
-        checkableItem = getOverlayMenuItems(getAllOverlays().get(1)).get(1);
+        checkableItem = menuBar
+                .getSubMenuItems(menuBar.getAllSubMenues().get(1)).get(1);
```
Comment (web-padawan) [inline vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java:112]:
```suggestion
                .getSubMenuItems(menuBar.getAllSubMenus().get(1)).get(1);
```

File: vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java [hunk 8]
```diff
@@ -81,42 +81,44 @@ public void clickRootButton_hoverOnParentItem_subSubMenuRenders() {
...
     }
 
     @Test
     public void setCheckedExternally_openSubMenu_itemChecked() {
         click("toggle-checked");
         openSubSubMenu();
-        TestBenchElement checkableItem = getOverlayMenuItems(
-                getAllOverlays().get(1)).get(1);
+        TestBenchElement checkableItem = menuBar
+                .getSubMenuItems(menuBar.getAllSubMenues().get(1)).get(1);
```
Comment (web-padawan) [inline vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow-integration-tests/src/test/java/com/vaadin/flow/component/menubar/tests/MenuBarPageIT.java:121]:
```suggestion
                .getSubMenuItems(menuBar.getAllSubMenus().get(1)).get(1);
```

File: vaadin-menu-bar-flow-parent/vaadin-menu-bar-testbench/src/main/java/com/vaadin/flow/component/menubar/testbench/MenuBarElement.java [hunk 9]
```diff
@@ -64,4 +70,55 @@ private boolean isVisible(TestBenchElement element) {
...
+    public List<TestBenchElement> getSubMenuItems(TestBenchElement overlay) {
+        return overlay.$("vaadin-context-menu-item").all();
+    }
+
+    /**
+     * Get the sub menu overlay element.
+     *
+     * @return TestBenchElement for the first open sub menu.
+     */
+    public TestBenchElement getSubMenu() {
```
Comment (web-padawan) [inline vaadin-menu-bar-flow-parent/vaadin-menu-bar-testbench/src/main/java/com/vaadin/flow/component/menubar/testbench/MenuBarElement.java:100]: I wonder if the name should be changed to e.g. `getSubMenuOverlay()` to make it clear?

There is `getSubMenu()` public API in the `MenuBar` that is used when creating items.
Some users might get confused by methods that are called the same, but with different meaning.

Comment (sissbruecker) [reply vaadin-menu-bar-flow-parent/vaadin-menu-bar-testbench/src/main/java/com/vaadin/flow/component/menubar/testbench/MenuBarElement.java:100]: I think it's fine as is, from a user perspective this element is effectively the "sub-menu", it being an overlay is more of a technical detail.

**Review by yuriy-fix — APPROVED** [review]

**Review by web-padawan — APPROVED** [review]

**Pull request merged**